Parser: Propose new hand-coded parser

dmsnell opened the pull request:
For some time we've needed a more performant PHP parser for the first
stage of parsing the `post_content` document.

See #1681 (early exploration)
See #8044 (parser performance issue)
See #1775 (parser performance, fixed in php-pegjs)

I'm proposing this implementation of the spec parser as an alternative
to the auto-generated parser from the PEG definition.

**Updates**
 - This now also includes a copy of the parser in JS whose performance is also quite good.
 - The files have been moved into the `/packages` directory - I still need some help understanding where it all belongs and how to make the package work

This provides a setup fixture for #6831 wherein we are testing alternate
parser implementations - https://comparator-yizlfvqafz.now.sh

## Distinctives
 - designed as a basic recursive-descent
 - but doesn't recurse on the call-stack, recurses via trampoline
 - moves linearly through document in one pass
 - relies on RegExp for tokenization

> **Note** I expect us to discover implementation bugs during the initial rollout of this parser. We have run it through our document library and unit tests but real posts are surely getting into more complicated constructions. We can deal with these as they come but we should expect these.

## Todo
 - [x] nested blocks include the nested content in their `innerHTML`
   this needs to go away
 - [x] create test fixture - https://comparator-yizlfvqafz.now.sh
 - [x] figure out where to save this file
 - [x] phpunit tests

## Benchmark

For posterity's sake I ran the merged parser through the parser comparator and compared it against the auto-generated spec parser. Here are the results from my laptop

```
                                    ms                        MB    
                                Spec  Default  Speedup    Spec  Default  Comparison
demo-post.html                    29.58   0.23   130     38.56   16.43     43%
early-adopting-the-future.html   263.83   1.01   262     36.84   17.10     46%
moby-dick-parsed.html           5012.13  11.55   434     75.41   25.18     33%
pygmalian-raw-html.html          330.35   0.24  1366    116.72   16.90     14%
redesigning-chrome-desktop.html  211.42   1.22   173     37.22   16.51     44%
shortcode-shortcomings.html       71.28   0.36   198     34.07   16.98     50%
web-at-maximum-fps.html          161.35   0.87   186     33.12   16.32     49%
```

The tests were done on my late 2013 rMBP quad core 2.6 GHz laptop. According to the Intel Power Gadget the CPU was running at 3.6 GHz the entire time. Each document was parsed with each parser at least 47 times and the runs were at random and each run was randomly chosen to parse the document between one and five times in a row before returning the results. Runtime and memory use were measured inside a runner script running in Docker as described in the parser comparator.

dmsnell committed:
Parser: Propose new hand-coded PHP parser

For some time we've needed a more performant PHP parser for the first
stage of parsing the `post_content` document.

See #1681 (early exploration)
See #8044 (parser performance issue)
See #1775 (parser performance, fixed in php-pegjs)

I'm proposing this implementation of the spec parser as an alternative
to the auto-generated parser from the PEG definition.

This is not yet ready to go but I wanted to get the code in a branch
so I can iterate on it and garner early feedback.

This should eventually provide a setup fixture for #6831 wherein we
are testing alternate parser implementations.

 - designed as a basic recursive-descent
 - but doesn't recurse on the call-stack, recurses via trampoline
 - moves linearly through document in one pass
 - relies on RegExp for tokenization

 - nested blocks include the nested content in their `innerHTML`
   this needs to go away
 - create test fixutre
 - figure out where to save this file

dmsnell commented:
I'm pretty sure that the next steps from here involve pondering the data structure of the stack. We have enough working knowledge now to know what we need to track and how we can pop that from the stack to the output.

dmsnell committed:
Fix issue with containing the nested innerHTML

dmsnell committed:
Also handle newlines as whitespace

dmsnell committed:
Use classes for some static typing

dmsnell committed:
add type hints

dmsnell committed:
remove needless comment

dmsnell committed:
space where space is due

dmsnell committed:
meaningless rename

dmsnell committed:
remove needless function call

dmsnell committed:
harmonize with spec parser

dmsnell committed:
don't forget freeform HTML before blocks

dmsnell committed:
account for oddity in spec-parser

dmsnell committed:
add some polish, fix a thing

dmsnell committed:
comment it

dmsnell committed:
add JS version too

dmsnell committed:
Change `.` to `[^]` because `/s` isn't well supported in JS

The `s` flag on the RegExp object informs the engine to treat a
dot character as a class that includes the newline character.
Without it newlines aren't considered in the dot.

Since this flag is new to Javascript and not well supported in
different browsers I have removed it in favor of an explicit class
of characters that _does_ include the newline, namely the open
exclusion of `[^]` which permits all input characters.

Hat-top to @hywan for finding this.

dmsnell committed:
Move code into `/packages` directory, prepare for review

dmsnell committed:
take out names from RegExp pattern to not fail tests

dmsnell committed:
Fix bug in parser: store HTML soup in stack frames while parsing

Previously we were sending all "HTML soup" segments of HTML between
blocks to the output list before any blocks were processed. We should
have been tracking these segments during the parsing and only spit
them out when closing a block at the top level.

This change stores the index into the input document at which that
soup starts if it exists and then produces the freeform block when
adding a block to the output from the parse frame stack.

dmsnell committed:
fix whitespace

dmsnell committed:
fix oddity in spec

pento commented on packages/block-serialization-default-parser/README.md:
```diff
@@ -0,0 +1,107 @@
+# Block Serialization Default Parser
+
+This library contains the default block serialization parser implementations for
+WordPress documents. It provides native PHP and Javascript parsers that implement
```
`s/Javascript/JavaScript/` 🙂

dmsnell replied:
substituted in ee72314cc

pento reviewed (commented):
Noice! Let's get this in sooner rather than later, so we can make inroads on the things depending on having a faster parser. 🙂

I've left some comments, here are a few random notes that have occurred to me, as well:
- It feels a little weird to be putting the PHP parser on NPM, but we don't really use Packagist at all, sooo... 🤷‍♂️  Let's stick with NPM for now, we can potentially explore doing Packagist/composer things later.
- `phpcs.xml.dist` needs to be updated to scan the new PHP code. I mentioned a couple of coding standards issues in the comments, but PHPCS should pick up the rest.
- Combined with switching the parser in `gutenberg_parse_blocks()`, `phpunit/class-parsing-test.php` should be updated to use `gutenberg_parse_blocks()`, rather than `Gutenberg_PEG_Parser`.

With this performance improvement, it seems like we could change `do_blocks()` to parse the content, instead of using the dynamic blocks regex.

mtias commented on packages/block-serialization-default-parser/parser.php:
```diff
@@ -0,0 +1,260 @@
+    static $parser;
+
+    if ( ! isset( $parser ) ) {
+        $parser = new BSDP_Parser();
```
Agreed. `Block_Parser()`?

dmsnell replied:
mainly this is there to prevent namespace collisions. my hope is that a few PRs after this we'll have a filter choose the parser and obviously if we create two or more `Block_Parser()` classes we'll run into conflicts.

any thoughts on that? even with an encapsulating class we run into some issues here because I don't think we can create a class within a class. the only way around it otherwise I think is actual namespacing which isn't supported on older PHP version…

pento replied:
Realistically, is there going to be a completely new parser appear between now and 5.0? It seems like this parser is going to be the one that will go into Core.

If that's the case, we should just use a generic name. `WP_Block_Parser` will fit into the WordPress naming scheme.

dmsnell commented on packages/block-serialization-default-parser/parser.php:
```diff
@@ -0,0 +1,260 @@
+<?php
+
+function bsdp_parse($document ) {
```
see related [comment response](
https://github.com/WordPress/gutenberg/pull/8083#discussion_r21283666)9 below.

I'm having some trouble understanding what you wrote @pento. I _hope_ we create a filter to select the parsing function but won't that depend somewhat on having unique names for each possible parse functions?

dmsnell replied:
experimented in 064efa58d but I haven't tested it yet

for what it's worth I'd be more comfortable getting this parser in first before making the parser system pluggable just because of the scope of the changes

dmsnell committed:
match styles

dmsnell commented on packages/block-serialization-default-parser/README.md:
```diff
@@ -0,0 +1,107 @@
+# Block Serialization Default Parser
+
+This library contains the default block serialization parser implementations for
```
this makes me want to cry since it's something I love about markdown and consistent among every other markdown parser I've used.

nonetheless, I have destroyed my markdown to make it happy in ee72314cc

🙃

dmsnell commented on packages/block-serialization-default-parser/parser.php:
```diff
@@ -0,0 +1,260 @@
+
+        switch ( $token_type ) {
+            case 'no-more-tokens':
+                # if not in a block then flush output
```
double-slashed it in ee72314cc

dmsnell commented on packages/block-serialization-default-parser/parser.php:
```diff
@@ -0,0 +1,260 @@
+                    return false;
+                }
+
+                # Otherwise we have a problem
```
exploded comments in ee72314cc

dmsnell committed:
use class name filter on server-side parser class

dmsnell committed:
fix whitespace

gziolo commented on packages/block-serialization-default-parser/package.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "@wordpress/block-serialization-default-parser",
+  "version": "1.0.0",
```
I would put `1.0.0-rc.0` or something like that to allow Lerna to do its job - it always bumps version so it would try to do `1.0.1` release otherwise ...

gziolo commented on webpack.config.js:
```diff
@@ -88,6 +88,7 @@ const gutenbergPackages = [
 	'blob',
 	'blocks',
+	'block-serialization-default-parser',
 	'block-serialization-spec-parser',
```
Can we stop bundling the other one if we don't use it in Gutenberg anymore?

dmsnell replied:
a good question. I don't want to kill the PEG parser since that maintains the spec in a way no hand-written implementation can.

in my comparator PRs I'm trying to move towards a system that will automatically run the implementations against the specification in something like a CI job so that we can have our formal specification without worrying about the implementation diverging (for example, if someone makes a change to the implementation without changing the spec first)

that is, I think we want to keep the `spec-parser` wherever we need it - mainly I think we want to strip it from the default _load_ of Gutenberg but whether we _build_ it, what do you think?

gziolo replied:
The package with transpiled code is going to be there anyway. It's really up to you and how you want to use it. If you are fine with referencing it as a regular npm package then you don't need it. If you want to consume it as part of e2e test or something which requires all Gutenberg build files then you can leave it as is. I just wanted to raise the awareness.

gziolo commented on lib/client-assets.php:
```diff
@@ -369,6 +376,7 @@ function gutenberg_register_scripts_and_styles() {
 			'wp-autop',
 			'wp-blob',
+			'wp-block-serialization-default-parser',
 			'wp-block-serialization-spec-parser',
```
I think we no longer need to list `wp-block-serialization-spec-parser` as a dependency. In addition, we should stop registering it, too.

dmsnell replied:
agreed on this one but I wasn't entirely sure how we wanted this to work…

do we want Gutenberg to automatically replace the spec parser with the "default" one at boot through a filter or do we want the "default" to be the default?

I want the auto-generated parser to be available still, especially for things like diagnostics and exploration.

gziolo replied:
As commented above, it all depends on the way you want to use it. I don't have any strong opinions about it. We should just ensure we don't ship unused code to the end users.

mcsf replied:
Do we have a decision here?

dmsnell replied:
I left the spec parser registered but un-enqueued it in 66455b461

gziolo commented on lib/blocks.php:
```diff
@@ -66,8 +66,16 @@ function gutenberg_parse_blocks( $content ) {
+	 *
+	 * @param string $parser_class Name of block parser class
+	 */
+	$parser_class = apply_filters( 'block_parser_class', 'BDSP_Parser' );
```
We should document it in the extensibility docs. Probably, the main document would be the best fit: https://github.com/WordPress/gutenberg/blob/master/docs/extensibility.md.

mcsf replied:
This still reads `BDSP`. :)

dmsnell replied:
another great catch - fixed in 66455b461

gziolo commented on packages/blocks/src/api/parser.js:
```diff
@@ -378,6 +378,6 @@ const createParse = ( parseImplementation ) =>
  * @return {Array} Block list.
  */
-export const parseWithGrammar = createParse( grammarParse );
+export const parseWithGrammar = createParse( defaultParse );
```
Should we offer a filter for JS implementation, too?

dmsnell replied:
yes but I wasn't sure if this PR was the right one for it. that is, filtering out the PHP side seemed somewhat straightforward while filtering the JS side seemed more complicated since we have to take into account things like loading the parser bundles and making sure they are available before the editor loads

do you think we need to do it all here in this PR?

gziolo replied:
It's totally fine as its own PR, I just wanted to ensure we tackle both PHP and JS side of things.

dmsnell committed:
Document extensibility

gziolo commented on docs/extensibility/parser.md:
```diff
@@ -0,0 +1,37 @@
+    return 'EmptyParser';
+}
+
+add_filter( 'block_parser_class', select_empty_parser, 10, 1 );
```
I think we provide the name of the function as a string in other examples to ensure it works with PHP 5.2. We might also want to prefix the function name with the plugin name:

```php
add_filter( 'block_parser_class', `my_plugin_select_empty_parser`, 10, 1 );
```

dmsnell replied:
good catch! I never meant to leave out the string - just neglected it - updated in 96ecfb8c2

gziolo commented:
8c7e42c looks great, I left one comment which is a tiny thing that affects only PHP 5.2...

dmsnell committed:
fix typo in example code

mcsf commented on packages/block-serialization-default-parser/parser.php:
```diff
@@ -0,0 +1,272 @@
+}
+
+function bdsp_select_parser( $prev_parse_class ) {
+    return 'BSDP_Parser';
```
There's a typo at _BSDP_. Anyway, given that the `apply_filters` call in `gutenberg_parse_blocks` defaults to `'BDSP_Parser'`, we should remove this bit.

dmsnell replied:
good catch! Is removed the function in 9c85a60cf

mcsf commented:
I'm getting a tokenization bug while testing with a personal post. Digging…

mcsf commented on packages/block-serialization-default-parser/src/index.js:
```diff
@@ -0,0 +1,239 @@
+	const namespace = namespaceMatch || 'core/';
+	const name = namespace + nameMatch;
+	const hasAttrs = !! attrsMatch;
+	const attrs = hasAttrs ? JSON.parse( attrsMatch ) : null;
```
I know there's a performance hit with `try`, but we should play it safe with `JSON.parse`, or generally speaking make sure we can inform the user of bad input and recover (e.g. isolate bad blocks) as best as possible. Thoughts, @dmsnell?

aduth replied:
There's no longer the famed V8 deoptimization with `try` / `catch`

https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#2-unsupported-syntax
https://github.com/v8/v8/commit/9aac80f

dmsnell replied:
added the `try` in 9c85a60cf but left it out of the PHP since in PHP it already returns `null` on a failed parse

mcsf committed:
Push failing parsing test

mcsf commented:
@dmsnell: I've pushed a failing test for the parser. The gist of it is that I think the tokenizer is too greedy when looking for the end of an attributes group (`{"some":"json"}`). Thus, a document with two self-closing attribute-equipped blocks, not necessarily consecutive, breaks the parser:

```html
<!-- wp:block {"ref":313} /-->
<!-- wp:block {"ref":482} /-->
```

This makes the parser throw a syntax error in the `JSON.parse` call:

```
SyntaxError: Unexpected token / in JSON at position 19
```

We should guarantee handling of any bad JSON here, but that's not the real issue. The issue is in the tokenizer, as the following fragment was returned as a match for `attrsMatch`:

```
{\"ref\":313} /--><!-- wp:block {\"ref\":482}
```

Note that, in contrast, the following input is correctly parsed:

```
<!-- wp:block {"ref":313} -->
<!-- /wp:block -->
<!-- wp:block {"ref":482} /-->
```

* * *

I used the following debugger patch:

```diff
diff --git a/packages/block-serialization-default-parser/src/index.js b/packages/block-serialization-default-parser/src/index.js
index 9c1983f22..007edd2b5 100644
--- a/packages/block-serialization-default-parser/src/index.js
+++ b/packages/block-serialization-default-parser/src/index.js
@@ -172,7 +172,7 @@ function nextToken() {
 	const namespace = namespaceMatch || 'core/';
 	const name = namespace + nameMatch;
 	const hasAttrs = !! attrsMatch;
-	const attrs = hasAttrs ? JSON.parse( attrsMatch ) : null;
+	const attrs = hasAttrs ? safeParse( attrsMatch ) : null;
 
 	// This state isn't allowed
 	// This is an error
@@ -192,6 +192,17 @@ function nextToken() {
 	return [ 'block-opener', name, attrs, startedAt, length ];
 }
 
+function safeParse( json ) {
+	let r;
+	try {
+		r = JSON.parse( json );
+	} catch ( e ) {
+		console.error( `Input of length ${ json.length }`, json );
+		throw e;
+	}
+	return r;
+}
+
 function addFreeform( rawLength ) {
 	const length = rawLength ? rawLength : document.length - offset;
```

dmsnell committed:
fix lazy/greedy bug in parser regexp

dmsnell commented:
> the tokenizer is too greedy when looking for the end of an attributes group

excellent find @mcsf! you are right - I let in a greedy match when I had no reason to! that's been taken out by the addition of the `?` to make the `(?!-->).` group un-greedy as it should be. I'm embarrassed that I let it in but so glad you found it and added the failing tests!

un-greedy modifier added in 9c85a60cf

also I rebased the branch

mcsf committed:
Docs: Fix typos, links, tweak style.

mcsf commented:
Thanks for the fixes, @dmsnell! I just pushed 05ee910, a revision of the docs.

Concerning `parser.md`, my main impression is that part of it could be removed or be made more succinct if we link to the right resources for more on the language and parsing/serialization stack. Otherwise, I feel we're repeating ourselves and diluting our lang resources.

mcsf commented:
Slating this for 3.8. We should merge this right after the 3.7 release (today-ish) to make sure any oddities can be spotted in `master`.

mcsf commented on docs/extensibility/parser.md:
```diff
@@ -0,0 +1,36 @@
+
+## Server-side parser
+
+Plugins have access to the parser if they want to process posts in their structured form instead of a plain HTML-as-string representation. WordPress itself uses the server-side parser when rendering pages: it removes the block comment delimiters and it replaces dynamic blocks with their rendered content.
```
> WordPress itself uses the server-side parser when rendering pages: it removes the block comment delimiters and it replaces dynamic blocks with their rendered content.

We're not there yet, we should remove this line.

dmsnell replied:
not there? last time I checked we were. the server-side parser is currently a partial implementation right? I thought for sure that we both had access to the full parse on the server (for plugins) _and_ WordPress itself was parsing parts of structure to remove comments and render dynamic blocks - not a full tree, but top-level blocks

mcsf replied:
We do process dynamic blocks, but we don't use the server-side parser (meaning `gutenberg_parse_blocks`) in `do_blocks` since #4591. I suggested removing this line because it sounds like the same (full) parser is being used for dynamic blocks.

After the current PR, I'm sure we'll be able to switch to just using the new parser, and update the documentation here.

dmsnell replied:
removed in 2e4fc348a

mcsf commented on packages/block-serialization-default-parser/README.md:
````diff
@@ -0,0 +1,67 @@
+        "innerHTML": "<!--more-->"
+    }
+]
+```
````
I suggest a more readable example. `core/more` is pretty quirky, since it's wrapping a core magical string `<!--more-->`. Anything else, really, as long as it's short and sweet.

dmsnell replied:
here I was just copying the same example from the spec parser which was pretty minimal but in 66455b461 I made a bigger example

mcsf commented on lib/blocks.php:
```diff
@@ -66,8 +66,16 @@ function gutenberg_parse_blocks( $content ) {
+	/**
+	 * Filter to allow plugins to replace the server-side block parser
+	 *
+	 * @since 3.7
```
Marking so we don't forget to bump this to 3.8.

dmsnell replied:
bumped in 66455b461

mcsf reviewed (commented):
Hopefully the last bits of feedback. :)

dmsnell committed:
update from PR feedback

dmsnell committed:
trim docs

mcsf commented:
Thanks for the fixes! Pushed a rebase. Now that 3.7 has been released, let's get this merged ASAP. I'll circle back to this PR very soon and hopefully approve it. Other eyes are also welcome.

mcsf committed:
Load default block parser, replacing PEG-generated one

mcsf commented on packages/block-serialization-default-parser/parser.php:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    function add_freeform( $length = null ) {
+        $length = $length ?: strlen( $this->document ) - $this->offset;
```
The `?:` ternary shorthand was only [introduced](https://secure.php.net/manual/en/language.operators.comparison.php) in PHP 5.3, while we need to support 5.2.

mcsf replied:
Addressed in 80279bb99.

mcsf committed:
Expand `?:` shorthand for PHP 5.2 compat

mcsf commented:
I pushed 299ddf7 in order to load on the new parser class on the server, otherwise calling `gutenberg_parse_block` fails. This is a bit naïve, though: we probably don't need the parser most of the time, especially if the `block_parser_class` filter is being used. Can we get away with a `require_once` fit inside `gutenberg_parse_blocks` after a check for `$parser_class === 'BSDP_Parser'`?

Finally, we need phpunit tests for this package. They can mimic what we have in `phpunit/class-parsing-test.php` for the PEG parser.

dmsnell committed:
add fixtures test for default parser

dmsnell committed:
spaces to tabs

dmsnell committed:
could we need no assoc?

dmsnell committed:
fill out return array

dmsnell committed:
put that assoc back in there

dmsnell committed:
isometrize

pento commented on lib/blocks.php:
```diff
@@ -66,8 +66,16 @@ function gutenberg_parse_blocks( $content ) {
+	/**
+	 * Filter to allow plugins to replace the server-side block parser
+	 *
+	 * @since 3.8
```
`@since 3.8.0` 😉

pento commented:
Whee, I finally got back to this PR! Nice work @dmsnell, this is looking really good. Apart from the one little docblock fix, the only other change I had was the class name: unless we're going to magically come up with a new parser to replace this one between now and 5.0, let's name it `WP_Block_Parser`, so it fits into the WP naming scheme, and we can avoid yet another rename when merge comes.

dmsnell committed:
rename and add 0

pento commented:
Oh, one more thing: the classes, members and methods need docblocks. 🙂

jorgefilipecosta committed:
Conditionally include the parser class

jorgefilipecosta commented on lib/blocks.php:
```diff
@@ -66,8 +66,20 @@ function gutenberg_parse_blocks( $content ) {
+	$parser_class = apply_filters( 'block_parser_class', 'WP_Block_Parser' );
+	// Load default block parser for server-side parsing if the default parser class is being used.
+	if ( 'WP_Block_Parser' === $parser_class ) {
+		require_once dirname( __FILE__ ) . '/../packages/block-serialization-default-parser/parser.php';
```
I added this conditional load as suggested by @mcsf. I did some tests and this approach seems to work well.
I don't like the fact that we are requiring an external file in the scope of a function but in this case, it seems the best alternative. The file being included implements a class and in PHP they are global so this definition works even with the include inside the function.
If someone has a better idea for this I'm totally open to change the approach.

dmsnell replied:
thanks! I have no problem with that

jorgefilipecosta reviewed (commented):
I did a set of tests with complex nesting scenarios and everything went great. I think the parsers is working fine 👍
@dmsnell let me know if you need any help with adding the documentation to the classes or any other pending task that I'm missing.

dmsnell committed:
Add docblocks

jorgefilipecosta reviewed (approved):
Thank you for your effort in getting this PR to the finish line, I think we are ready to merge this 👍 
The core of the parser seems to work correctly and if later we find improvements/alternative approaches, we can iterate on other PR's.
Awesome work!

youknowriad commented on packages/block-serialization-default-parser/package.json:
```diff
@@ -0,0 +1,25 @@
+  "module": "build-module/index.js",
+  "publishConfig": {
+    "access": "public"
+  }
```
I think this needs a babel-runtime dependency cc @gziolo ?

gziolo replied:
I see the following output in the transpiled code:

```js
// ES5
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

// ESM
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
```

So it needs to be there.

gziolo replied:
Added in 9463906.

gziolo committed:
Standardize the package configuration

pento reviewed (approved):
It is time. Click the button.

gziolo commented:
💥
done
😅

mtias commented:
Thanks for all the work @dmsnell ! Nice to see this coming together.

aduth commented:
>This now also includes a copy of the parser in JS whose performance is also quite good.

Do we have numbers for this, as far as performance comparison to the previous parser?

mcsf commented:
Thanks, everyone, for seeing this through!

mcsf commented:
Concerning the requiring of the PHP implementation, #9791 needs investigating.

mcsf commented on packages/block-serialization-default-parser/parser.php:
```diff
@@ -0,0 +1,449 @@
+	 * @since 3.8.0
+	 * @var WP_Block_Parser_Block[]
+	 */
+    public $output;
```
I'm concerned about this promise that `$output` is an array of `WP_Block_Parser_Block`, since freeform fragments are added as [associative] arrays and not class instances.

dmsnell replied:
we can definitely consider wiping the output clean of its classes - I didn't at first because it seemed benign to retain them, but if we sacrifice a little performance we can `json_decode( json_encode( $output ) )` and clear it up

mcsf replied:
@jorgefilipecosta mentioned implementing an `ArrayObject` interface in our classes so that one can traverse our parser output natively, rather than doing the JSON dance. What do you think?

dmsnell replied:
That's a good question. It means more divide between the PHP and JS versions of the parser. What's the JSON dance? Wouldn't having `ArrayObject` be somewhat superfluous?

```php
// this already works with arrays and objects!
$blocks = parse( $document );
$blocks = array_map( $blocks, $my_transformer );
```

we probably want to fix the bug as a separate thing from adding interfaces. I'm skeptical of the value of the latter if the former is resolved.

mcsf replied:
By JSON dance I meant `json_decode( json_encode( $output ) )`, sorry for not being clear.

> we probably want to fix the bug as a separate thing from adding interfaces

So this is the actual issue: #10047. It's not the traversal (looking at your `array_map`) example but rather accessing properties of a block, which can either mean accessing properties of an array or of an object.

jorgefilipecosta replied:
Classes offer some advantages we can publish abstract classes that contain the fields plugins can safely access, and other parsers can just extend this general classes. Simple arrays don't offer this guarantees.

But now we have a problem some plugins are dependent on using simple arrays, even if this bug was already caught I'm not certain we can just change the API to use classes.

So I think our options are revert back and just use arrays, or advance and change our API to use classes, but in this case in order to be back-compatible with existing implementations accessing using the array syntax I think our only solution is ArrayObject which allows us to temporarily return something that behaves like a class for new implementations and an array for old implementations, in this case, we would add the deprecation messages saying we now return objects.

dmsnell replied:
> It's not the traversal (looking at your array_map example) but rather accessing properties of a block, which can either mean accessing properties of an array or of an object.

to me this is just evidence that the work to make all attribute reporting consistent is necessary. some attributes are `null`, some are objects

> By JSON dance I meant json_decode( json_encode( $output ) ), sorry for not being clear.

that would be in the parser and wouldn't have to be manually performed. in fact, the classes are only even there for performance, so we can test the change of sorting everything in plain old objects vs. converting at the end. if it's a degradation then we can simply remove the classes if we want to preserve the simpler interface.

mcsf reviewed (commented):
I hadn't realized this before — as my primary testing interface was the WP API ([gist](https://gist.github.com/mcsf/1f67c86a5e40dbbf630def4625348b7c)), through which everything is serialized into the same shape — but I now fear that we're not providing a consistent interface with the parser in its current state.

See my inline comments. Consumers of `gutenberg_parse_blocks` may make mistakes because of these discrepancies, and I fear they may already have: #10041.

cc @dmsnell

dmsnell commented on packages/block-serialization-default-parser/parser.php:
```diff
@@ -0,0 +1,449 @@
+
+        if ( isset( $stack_top->leading_html_start ) ) {
+            $this->output[] = array(
+                'attrs' => array(),
```
good call - I know there are some lingering inconsistencies too around `null` vs. `{}` in the spec grammar. a good follow-up PR that's been on my TODO list
